cdk/domino_cdk/provisioners/eks/eks_nodegroup.py

```diff
@@ -0,0 +1,278 @@
+                            ),
+                        )
+                    ],
+                    role=self.ng_role,
+                    instance_type=ec2.InstanceType(ng.instance_types[0]),
+                    key_name=ng.key_name,
+                    machine_image=machine_image,
+                    user_data=mime_user_data,
+                    security_group=self.unmanaged_sg,
+                )
+                # mimic adding the security group via the ASG during connect_auto_scaling_group_capacity
+                lt.connections.add_security_group(self.cluster.cluster_security_group)
+                cfn_lt: ec2.CfnLaunchTemplate = lt.node.default_child
+                lt_data = ec2.CfnLaunchTemplate.LaunchTemplateDataProperty(
+                    **cfn_lt.launch_template_data._values,
+                    metadata_options=ec2.CfnLaunchTemplate.MetadataOptionsProperty(
+                        http_endpoint="enabled", http_tokens="required", http_put_response_hop_limit=2
+                    ),
+                )
+                cfn_lt.launch_template_data = lt_data
+
+            # https://github.com/aws/aws-cdk/issues/6734
+            cfn_asg: aws_autoscaling.CfnAutoScalingGroup = asg.node.default_child
+            # Remove the launch config from our stack
+            asg.node.try_remove_child("LaunchConfig")
+            cfn_asg.launch_configuration_name = None
+            # Attach the launch template to the auto scaling group
+            cfn_asg.mixed_instances_policy = cfn_asg.MixedInstancesPolicyProperty(
+                launch_template=cfn_asg.LaunchTemplateProperty(
+                    launch_template_specification=cfn_asg.LaunchTemplateSpecificationProperty(
+                        launch_template_id=cfn_lt.ref,
+                        version=lt.version_number,
+                    ),
+                    overrides=[cfn_asg.LaunchTemplateOverridesProperty(instance_type=it) for it in ng.instance_types],
+                ),
+            )
+
+            options: dict[str, Any] = {
+                "bootstrap_enabled": ami_id is None,
+            }
+            if not ami_id:
+                extra_args: list[str] = []
+                if labels := ng.labels:
+                    extra_args.append(
+                        "--node-labels={}".format(",".join(["{}={}".format(k, v) for k, v in labels.items()]))
+                    )
+
+                if taints := ng.taints:
+                    extra_args.append(
+                        "--register-with-taints={}".format(",".join(["{}={}".format(k, v) for k, v in taints.items()]))
+                    )
+                options["bootstrap_options"] = eks.BootstrapOptions(kubelet_extra_args=" ".join(extra_args))
+
+            self.cluster.connect_auto_scaling_group_capacity(asg, **options)
+
+    def _get_machine_image(self, cfg_name: str, image: MachineImage) -> Tuple[Optional[str], Optional[str]]:
```

Comment:
I do not welcome this change to passing MachineImage instead of config.

Default EKS AMI doesn't have meaningful user_data. EKS adds user_data directly to the launch template (via nodegroup). Adding arbitrary user_data to machine image object and then extracting this user_data from machine image just to add it to a LaunchTemplate is counterproductive and subject to side effects.

I recommend to keep ami_id and user_data separate until they are combined into a launch template.

Currently this code breaks user_data handling because machine_image is created without user_data (in provision_xxx_nodegroup second line) and here we are getting zilch.

Reply:
There might be a misunderstanding here, the `MachineImage` object being passed in here is the object from the config, _not_ a CDK machine image object. Maybe I should do some followup work to disambiguate/clearly namespace our config objects from aws api calls/results.

Reply:
Weird, I commented on this, it showed up twice, I deleted the second iteration of the comment, reloaded, and now it's gone entirely. Lovely. Guessing the two comments was just a display error. Anyway...

I think there might be a misunderstanding, the `image` variable being passed here's type, `MachineImage`, is the domino_cdk config object (ie just a container for the config), and *NOT* an aws cdk representation of a machine image. So we are still passing the config, but we're just passing the relevant portion of the config.

I think maybe I need to do some followup work to the object-based config to disambiguate/namespace the config objects so they don't get confused with aws_cdk things (like, as imported, `EKS` is a config object but `eks` is the aws_cdk's eks api... which is confusing). I will get that into my next PR.

Reply:
Agree. Confused with IMachineImage

Reply:
See my previous comment.

How about eliminating MachineImage and _get_machine_image (it doesn't do much anymore).

user_data is launch template concept, not an image concept.

Reply:
I was already planning to get rid of `_get_machine_image` in followup, so I will handle all three things (getting rid of the machine image config block, removing _get_machine_image, and disambiguating the config variables) in the followup.